src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,269 @@
+
+            // Start message queue processor
+            _cancellationTokenSource = new CancellationTokenSource();
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public void EnqueueMessage(string message)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                while (_messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                }
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch, _cancellationTokenSource);
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.WriteMessagesFailed(_logger, ex);
+                    }
+
+                    _currentBatch.Clear();
+                }
+                else
+                {
+                    try
+                    {
+                        await Task.Delay(_flushInterval, _cancellationTokenSource.Token);
+                    }
+                    catch
+                    {
+                        // Exit if task was canceled
+                        return;
+                    }
+                }
+            }
+        }
+
+        private async Task WriteMessagesAsync(List<string> messages, CancellationTokenSource cancellationTokenSource)
+        {
+            // Files are grouped by day, and written up to _maxFileSize before rolling to a new file
+            DateTime today = DateTime.Now;
+            var fullName = GetFullName(today);
+            var fileInfo = new FileInfo(fullName);
+            if (!TryCreateDirectory())
+            {
+                // return early if we fail to create the directory
+                return;
+            }
+            var streamWriter = GetStreamWriter(fullName);
+
+            try
+            {
+                foreach (var message in messages)
+                {
+                    if (cancellationTokenSource.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                    fileInfo.Refresh();
+                    // Roll to new file if _maxFileSize is reached
+                    // _maxFileSize could be less than the length of the file header - in that case we still write the first log message before rolling
+                    if (fileInfo.Exists && fileInfo.Length > _maxFileSize)
+                    {
+                        _fileNumber++;
```

Comment:
What happens if the app restarts and we've already filled multiple files for the day? `_fileNumber` would be reset to 0, right? And it looks like we only increment at most one file per message when the file is full. Does this just append one message to each existing file?

Reply:
If I'm right about this, I'm thinking it's okay to have a tight loop here that keeps checking whether the next file is full and increments `_fileNumber` without attempting to write a message.

We'll have to be careful to ensure that we actually overwrite the files when we loop back to zero and don't keep appending or infinitely loop.

Reply:
I'm appending a timestamp to the folder names now, which avoids this issue, but it does mean that the app doesn't write to the same folder after restart, and rolling doesn't delete pre-restart files. I'll fix this to remove the timestamp and implement what you suggested here, plus the following from Teams:

> make sure there's some logic that if you find 10,000 files in a row that are full, you log a scary/critical error and stop looping and give up on writing the message